packages/@cdktf/provider-schema/src/cache.ts

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import { ConstructsMakerTarget, logger } from "@cdktf/commons";
+import * as fs from "fs-extra";
+
+// We keep this very simple since the caching feature is experimental
+// we might need to do housekeeping / include terraform / cdktf version in the future
+function cacheKey(input: ConstructsMakerTarget): string {
+  return `${encodeURIComponent(input.fqn)}@${input.version}`;
+}
+
+export function cachedAccess<I extends ConstructsMakerTarget, O>(
+  producer: (input: I) => Promise<O>,
+  cacheDir?: string | null
+): (input: I) => Promise<O> {
+  const cacheEnabled = Boolean(
+    cacheDir && fs.lstatSync(cacheDir).isDirectory()
```

Comment:
I'd rather error if a dir was passed but is not a valid directory. Makes it easier to spot why caching is not happening.

Reply:
Yeah good point, silently failing is never good